### ISSUE TITLE
Add to docs: Using with PrimeFaces

### ISSUE THREAD

**Issue by Vyling** [mkoryak/floatThead#378]
Hey, it's me again. I said it'd work fine now, but... I stumbled upon the next issue and finally got a working jsfiddle for that. For the scenario:
Got a table with responsive container, fixed width, fixed header which scrolls with the page. 
However, together with the fixed header, the table can not be resized under its current size anymore. See the following jsfiddle:
http://jsfiddle.net/fYp44/47/
Making it bigger doesn't seem to be a problem. But under certain circumstances, I need to scale it down. Would be great if it'd be supported.

Again, latest library version, Firefox 54.

**Comment by mkoryak** [comment]
You must always reflow when changing any Dom dimensions

http://jsfiddle.net/fYp44/49/

**Comment by Vyling** [comment]
I see, totally forgot that there. Sorry about that. 
When I do that with PrimeFaces' expanding resize mechanism, it will resize the table up and down along the columns to a certain point... but at a certain point it will only resize the columns, but not the table anymore....
I will experiment further whether it's an issue with with PrimeFaces or with the plugin. Maybe just a selector that messes up due to the added floatThead. If I can find out anything more certain I will try to do another jsfiddle. I wish I could do one with PrimeFaces' datatables, that would make it easier (also for you to say whether it's supported or not), but unfortunately, the source is not easily embeddable.
Sorry for the rambling, just trying to explain my situation in case you might be interested.

**Comment by mkoryak** [comment]
Ah, I did JSF back in the day, PrimeFaces sound related. If its anything like JSF from 7 years ago, then you are doing gods work :)

you might want to check if `position:'absolute'` helps you. It might result is slightly lower performance but make your life considerably easier - that positioning system requires less reflows, at least for position changes on the Y axis. This is because the table is wrapped by a container and the floated header lives at the top of that container. The downside of course is that now your table is wrapped, so some selectors will break. So maybe this is not a good option for you since you want to make a tool that works "anywhere"

also be careful with making sure that the number of columns in your header match with the number of columns in the body. Your last fiddle had that problem.

I am happy to help in any way, since it sounds like you are solving an interesting problem. If you cant make a fiddle but can put running (ideally clean) looking example somewhere on the internet, I would be happy to take a look

**Comment by mkoryak** [comment]
another thing you might try is `autoReflow: true` though it uses MutationObserver which is known to be slow on mobile, and doesnt work in some crappy IEs, and also I have not tested it much here

**Comment by Vyling** [comment]
Yeah that fiddle was messy, sorry. Since the table is generated by JSF its columns should match.
Unfortunately I can't rely on autoReflow, since I need to support IE8+. But even with it, the result was the same.
But I managed to solve it on my own.

For anyone else having problems with the Primefaces dataTable resizing:
I overwrote "PrimeFaces.widget.DataTable.prototype.resize". I copied its current implementation from here: https://jar-download.com/explore-java-source-code-detail.php?file=./javaSource/org.primefaces/primefaces/6.0/META-INF/resources/primefaces/datatable/datatable.js&key=a1e77aa03ca4a6dcabdbd8a13a9e5b97 (under "resize:")
What I noticed is that, on one hand the min-width-attribute was checked against by the PrimeFaces implementation. Since the stickyHeader adds that attribute, the resize wasn't performed properly when the size would go beneath that. So I overwrote it in CSS with 0px!important. Now it at least performed the necessary methods, yet it still didn't go under the min-width size.
I changed the "table" selector to perform on the underlying table (this.thead.closest(".ui-datatable-tablewrapper").find("table");), so it doesn't perform the changes on the floatThead-table-header, and now it seems to work perfectly fine. (At least with 'fixed'. With 'absolute' I get a floatThead-table with a colgroup, but without thead...)

So you might want to add this to your docs for people who might be looking for solutions with PrimeFaces. :)

Now I only have some issues with reordering... which I didn't have before, reflow worked wonders until now, so I guess it will be just another bug in combination with PrimeFaces. Now the column disappears into the void. Which is weird, because my change shouldn't have affected draggable columns.
But so far these problems have been far easier to figure out than the issues I had with PrimeFaces' own stickyHeader implementation.

**Comment by Vyling** [comment]
Alright, got it to work by just initializing the floatThead on page load for each table and triggering reflow onTabShow.

The only issue I have left now is the previous one, where you've written "Does not support multiple scroll containers controlling X and Y". It... kind of sounds like the following issue, but I'm not sure, as it refers to my old ticket with only one table and now  I only got that issue with multiple tables (hidden behind tabs).
 I created a jsfiddle which shows faulty behavior with multiple tables and that setup... http://jsfiddle.net/fYp44/50/ I am not sure if this problem depicted in the jsfiddle is related to my own.
In conclusion... everything works fine on the first table... However the second table won't have the header within the container anymore (because the floatThead-Container has the width of the table. But even when I shrink it, I will only see that shrinked part moved to left and right, without the rest of the header showing up as it should, or parts disappearing as they should).
I am not sure if I should open another issue for this one...

**Comment by Vyling** [comment]
JSF is still a popular framework, although opinions vary greatly, but it is still used quite frequently. At least that's my impression when searching for issues.

Btw, your donation link doesn't work.

**Comment by Vyling** [comment]
I got it to work. By changing the responsive container declaration as done here: http://jsfiddle.net/fYp44/51/
The problem was that he always took the first container. Which was hidden when changing tabs. So I instructed him to take the closest to the current context. I am not sure if it could be solved in the library or if the support for multiple tables would have to go this way then, but that is how it works for me now. Even with multiple visible tables on one page.

**Comment by mkoryak** [comment]
What you have there with finding the closest container is the standard pattern, I don't think its too much to ask my users to use the closest selector. I try to make my plugin do the least amount of magic possible to do its work.

**Comment by mkoryak** [comment]
I am happy to add something to the docs if you can provide me with the javascript you wrote to override the primefaces stuff, and even any java or jsp.

**Comment by mkoryak** [comment]
Yes, my donate link is not working. If you want to donate using paypal my paypal email is koryak.other@gmail.com

**Comment by Vyling** [comment]
This is the slightly edited Javascript. (not tested with ColumnGroups, neither scrollable, nor liveResize, but with expandMode = expand)
```
PrimeFaces.widget.DataTable.prototype.resize = function(event, ui) {
        var columnHeader, nextColumnHeader, change = null, newWidth = null, nextColumnWidth = null, 
        expandMode = (this.cfg.resizeMode === 'expand'),
        table = this.thead.closest(".ui-datatable-tablewrapper").find("table");
                
        if(this.hasColumnGroup) {
            var groupResizer = this.findGroupResizer(ui);
            if(!groupResizer) {
                return;
            }
            
            columnHeader = groupResizer.parent();
        }
        else {
            columnHeader = ui.helper.parent();
        }
        
        var nextColumnHeader = columnHeader.nextAll(':visible:first');
        
        if(this.cfg.liveResize) {
            change = columnHeader.outerWidth() - (event.pageX - columnHeader.offset().left),
            newWidth = (columnHeader.width() - change),
            nextColumnWidth = (nextColumnHeader.width() + change);
        }
        else {
            change = (ui.position.left - ui.originalPosition.left),
            newWidth = (columnHeader.width() + change),
            nextColumnWidth = (nextColumnHeader.width() - change);
        }
         
        var minWidth = parseInt(columnHeader.css('min-width'));
        minWidth = (minWidth == 0) ? 15 : minWidth;
        
        if((newWidth > minWidth && nextColumnWidth > minWidth) || (expandMode && newWidth > minWidth)) {
            if(expandMode) {
                table.width(table.width() + change);
                setTimeout(function() {
                    columnHeader.width(newWidth);
                }, 1);
            }
            else {
                columnHeader.width(newWidth);
                nextColumnHeader.width(nextColumnWidth);
            }
            
            if(this.cfg.scrollable) {
                var cloneTable = this.theadClone.parent(),
                colIndex = columnHeader.index();
        
                if(expandMode) {
                    var $this = this;
                    
                    //body
                    cloneTable.width(cloneTable.width() + change);
                    
                    //footer
                    this.footerTable.width(this.footerTable.width() + change);
                    
                    setTimeout(function() {
                        if($this.hasColumnGroup) {
                            $this.theadClone.find('> tr:first').children('th').eq(colIndex).width(newWidth);            //body
                            $this.footerTable.find('> tfoot > tr:first').children('th').eq(colIndex).width(newWidth);   //footer
                        }
                        else {
                            $this.theadClone.find(PrimeFaces.escapeClientId(columnHeader.attr('id') + '_clone')).width(newWidth);   //body
                            $this.footerCols.eq(colIndex).width(newWidth);                                                          //footer
                        }
                    }, 1);
                }
                else {
                    if(this.hasColumnGroup) {
                        //body
                        this.theadClone.find('> tr:first').children('th').eq(colIndex).width(newWidth);
                        this.theadClone.find('> tr:first').children('th').eq(colIndex + 1).width(nextColumnWidth);
                        
                        //footer
                        this.footerTable.find('> tfoot > tr:first').children('th').eq(colIndex).width(newWidth);
                        this.footerTable.find('> tfoot > tr:first').children('th').eq(colIndex + 1).width(nextColumnWidth);
                    }
                    else {
                        //body
                        this.theadClone.find(PrimeFaces.escapeClientId(columnHeader.attr('id') + '_clone')).width(newWidth);
                        this.theadClone.find(PrimeFaces.escapeClientId(nextColumnHeader.attr('id') + '_clone')).width(nextColumnWidth);
                        
                        //footer
                        if(this.footerCols.length > 0) {
                            var footerCol = this.footerCols.eq(colIndex),
                            nextFooterCol = footerCol.next();

                            footerCol.width(newWidth);
                            nextFooterCol.width(nextColumnWidth);
                        }
                    }
                }
            }            
        }
    }
```
The JSF bit within p:dataTable (which is within a tabview): 
```
<p:ajax event="colResize" onstart="$('div.ui-tabs-panel:visible').find('table').trigger('reflow');" />
```
I still need to figure out how to call the listener together with the colResize-Event-Object *after* reflow was executed. Currently I get the old width data in the event object. 
But for the UI, this does the job.

**Comment by mkoryak** [comment]
if you need to do stuff post reflow, look at the docs for the `reflowed` event. it should help

**Comment by Vyling** [comment]
Unfortunately it doesn't help, because I need to invoke a server side listener with an object which is only available in that tag. I can't use Javascript-code in the listener-attribute. I fear that's a more JSF-specific problem. If I wouldn't need that object, I could just setup a remoteCommand which is called oncomplete and does the thing, but I can't pass the object through Javascript... I guess there's some trick I can use with remoteCommand and passing widths along with it, instead of using the object, but that will need some experimenting.

**Comment by Vyling** [comment]
I tried the following...
```
var _self = this; 
$('div.ui-tabs-panel:visible').find('table').on('reflowed', function(e, $floatContainer{
        console.log(JSON.stringify(_self.data));
}); 
$('div.ui-tabs-panel:visible').find('table').trigger('reflow');
```
... and it keeps returning 'undefined' in the console, whereas "this.data" outside the event would return a string with various data, such as the id and width of the changed column. Maybe you can help me to set that up correctly?

**Comment by mkoryak** [comment]
if its undefined that means that the event fired before the data was set,
or it was unset before the event fired. I cant help you set it up correctly
because I would need to debug it to see what the problem is.

Perhaps if you `console.log(JSON.stringify(_self.data));` right after the
line where you set `_self` you will see that it is also undefined there

**Comment by Vyling** [comment]
Whoops, indeed. It's not defined onstart. I put it oncomplete, but it still returns the same wrong value within the event.

I'm going to work around that now by simply reading the table and each column's width.

**Comment by Vyling** [comment]
Alright... the following finally works.
In the DataTable:
```
<p:ajax event="colResize" onstart="var activeTable = $('div.ui-tabs-panel:visible').find('table'); activeTable.on('reflowed', function(e, $floatContainer){getColumnWidths();}); activeTable.trigger('reflow');"
				oncomplete="$('div.ui-tabs-panel:visible').find('table').off('reflowed')" />
```

in seperate js-file:
```
function getColumnWidths() {
        	var currentDataTableHeaders = $('div.ui-tabs-panel:visible').find('table th');
        	currentDataTableHeaders.each(function() {
        		saveColumnWidth([{name:'column', value:$(this).text()}, {name:'width', value:$(this).outerWidth()}]);
        	});
        }
```

below the DataTable:
```
<p:remoteCommand name="saveColumnWidth" process="@this"
			actionListener="#{bean.saveColumnWidth(param.column, param.width)}" />
```

So... that is how to get correct widths for each column. Couldn't figure out a better way.

**Comment by mkoryak** [comment]
note to self: 

add this to the docs

**Comment by mkoryak** [comment]
I finally did it! Its in the docs now:

http://mkoryak.github.io/floatThead/integrations/
